vault/seal_autoseal.go

```diff
@@ -499,3 +513,84 @@ func (d *autoSeal) migrateRecoveryConfig(ctx context.Context) error {
 
 	return nil
 }
+
+// StartHealthCheck starts a goroutine that tests the health of the auto-unseal backend once every 10 minutes.
+// If unhealthy, logs a warning on the condition and begins testing every one minute until healthy again.
+func (d *autoSeal) StartHealthCheck() {
+	d.StopHealthCheck()
+	d.hcLock.Lock()
+	defer d.hcLock.Unlock()
+
+	healthCheck := time.NewTicker(sealHealthTestIntervalNominal)
+	d.healthCheckStop = make(chan struct{})
+	healthCheckStop := d.healthCheckStop
+
+	go func() {
+		ctx := d.core.activeContext
+		lastTestOk := true
+		lastSeenOk := time.Now()
+		for {
+			select {
+			case <-healthCheckStop:
+				if healthCheck != nil {
+					healthCheck.Stop()
+				}
+				healthCheckStop = nil
+				return
+			case t := <-healthCheck.C:
+				func() {
+					ctx, cancel := context.WithTimeout(ctx, sealHealthTestTimeout)
+					defer cancel()
+
+					testVal := fmt.Sprintf("Heartbeat %d", mathrand.Intn(1000))
+					ciphertext, err := d.Access.Encrypt(ctx, []byte(testVal), nil)
+
+					if err != nil {
+						d.logger.Warn("failed to encrypt seal health test value, seal backend may be unreachable", "error", err)
+						if lastTestOk {
+							healthCheck.Reset(sealHealthTestIntervalUnhealthy)
+						}
+						lastTestOk = false
+						d.core.MetricSink().SetGauge(autoSealUnavailableDuration, 0)
+					} else {
+						plaintext, err := d.Access.Decrypt(ctx, ciphertext, nil)
+						if err != nil {
+							d.logger.Warn("failed to decrypt seal health test value, seal backend may be unreachable", "error", err)
+							if lastTestOk {
+								healthCheck.Reset(sealHealthTestIntervalUnhealthy)
+							}
+							lastTestOk = false
+						}
+						if !bytes.Equal([]byte(testVal), plaintext) {
+							d.logger.Warn("seal health test value failed to decrypt to expected value")
+							if lastTestOk {
+								healthCheck.Reset(sealHealthTestIntervalUnhealthy)
+							}
+							lastTestOk = false
+						} else {
+							d.logger.Debug("seal health test passed")
+							if !lastTestOk {
+								d.logger.Info("seal backend is now healthy again", "downtime", t.Sub(lastSeenOk).String())
+							}
+							if !lastTestOk {
+								healthCheck.Reset(sealHealthTestIntervalNominal)
+							}
+							lastTestOk = true
+							lastSeenOk = t
+						}
+						d.core.MetricSink().SetGauge(autoSealUnavailableDuration, float32(time.Since(lastSeenOk).Nanoseconds()))
```

Comment:
We tend to use millis for our time-based metrics, any reason you want nanos here?